<a href="https://colab.research.google.com/github/HSaleem352/credit-risk-classification/blob/main/credit_risk_classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.metrics import balanced_accuracy_score
from imblearn.metrics import classification_report_imbalanced

import warnings
warnings.filterwarnings('ignore')

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [4]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data_df = pd.read_csv(
    Path('Resources/lending_data.csv')
)

# Review the DataFrame
lending_data_df.sample(10)

loan_size  interest_rate  borrower_income  debt_to_income  \
23259     9700.0          7.267            49000        0.387755   
67587    10500.0          7.605            52200        0.425287   
56834     9400.0          7.139            47800        0.372385   
65141    11700.0          8.077            56600        0.469965   
16709     9300.0          7.056            47000        0.361702   
60762     9300.0          7.098            47400        0.367089   
34682     7700.0          6.410            40900        0.266504   
12702    10400.0          7.551            51700        0.419729   
73522     7600.0          6.367            40500        0.259259   
43790    10500.0          7.605            52200        0.425287   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
23259                4                 0       19000            0  
67587                4                 1       22200            0  
56834                4                 0       17800            0  
65141                5                 1       26600            0  
16709                3                 0       17000            0  
60762                3                 0       17400            0  
34682                2                 0       10900            0  
12702                4                 1       21700            0  
73522                2                 0       10500            0  
43790                4                 1       22200            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [8]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_data_df['loan_status']

# Separate the X variable, the features
x = lending_data_df.drop(columns = 'loan_status')

In [6]:
# Review the y variable Series
y.sample(10)

10818    0
16036    0
10999    0
24547    0
10780    0
73782    0
69198    0
12587    0
25704    0
23293    0
Name: loan_status, dtype: int64

In [9]:
# Review the X variable DataFrame
x.sample(10)

loan_size  interest_rate  borrower_income  debt_to_income  \
38130     9000.0          6.937            45900        0.346405   
20234    12000.0          8.230            58000        0.482759   
64725     9000.0          6.969            46200        0.350649   
28202    10400.0          7.563            51800        0.420849   
40037     9400.0          7.105            47500        0.368421   
40556     9100.0          6.988            46400        0.353448   
30149    10400.0          7.545            51600        0.418605   
37756     7700.0          6.385            40700        0.262899   
2949     10500.0          7.608            52200        0.425287   
76897    20000.0         11.640            90100        0.667037   

       num_of_accounts  derogatory_marks  total_debt  
38130                3                 0       15900  
20234                6                 1       28000  
64725                3                 0       16200  
28202                4                 1       21800  
40037                3                 0       17500  
40556                3                 0       16400  
30149                4                 1       21600  
37756                2                 0       10700  
2949                 4                 1       22200  
76897               13                 3       60100

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [11]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
x_train, x_test, y_train, y_test = train_test_split(
    x,
    y,
    random_state = 1
)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [12]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
LR_model = LogisticRegression(random_state = 1)

# Fit the model using training data
LR_model.fit(x_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [13]:
# Make a prediction using the testing data
LR_predictions = LR_model.predict(x_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [14]:
# Generate a confusion matrix for the model
'''
-- Out of the 18,765 loan status's that are healthy, the model predicted 18,663 as
   healthy correctly and 102 as healthy incorrectly

-- Out of the 619 loan status's that are non-healthy (high-risk), the model
   predicted 563 as non-healthy correctly and 56 as non-healthy incorrectly
'''

cm_imbalanced = confusion_matrix(y_test, LR_predictions)
cm_imbalanced_df = pd.DataFrame(cm_imbalanced,
                                index = ['Actual Healthy Loans (low-risk)',
                                'Actual Non-Healthy Loans (high-risk)'],
                                columns = ['Predicted Healthy Loans (low-risk)', 'Predicted Non-Healthy Loans (high-risk)']
                              )
cm_imbalanced_df

Predicted Healthy Loans (low-risk)  \
Actual Healthy Loans (low-risk)                                    18663   
Actual Non-Healthy Loans (high-risk)                                  56   

                                      Predicted Non-Healthy Loans (high-risk)  
Actual Healthy Loans (low-risk)                                           102  
Actual Non-Healthy Loans (high-risk)                                      563

In [15]:
# Print the classification report for the model
print(classification_report_imbalanced(y_test, LR_predictions))

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.91      1.00      0.95      0.91     18765
          1       0.85      0.91      0.99      0.88      0.95      0.90       619

avg / total       0.99      0.99      0.91      0.99      0.95      0.91     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:**
- The model seems to be doing great based on its balanced accuracy score of 95%. But this high score is mostly because the data is skewed. There are way more healthy loans (low-risk) than non-healthy ones (high-risk). This means the model is better at predicting loans as healthy rather than identifying them as non-healthy.

- Based on the imbalanced classification report, the model always predicted healthy loans correctly 100% of the time, but it only got non-healthy loans right 85% of the time.

---